### PR TITLE
Test with minimal dependency versions

### LINKED ISSUE #1230: `Cargo.lock` strategy
The current recommendation is to not commit `Cargo.lock` for libraries but commit it for binaries. The main reason being that it doesn't affect downstream crates. However I think that committing it has some value - it says "we tested this crate with these versions of dependencies". And if we review the crates (ensuring the hashes match) it also acts as a lightweight `cargo-crev`. (I myself don't want to commit to this for now, maybe around 1.0 where versions will be more stable.)

However we could have these problems: people accidentally changing and committing it possibly causing merge conflicts, older versions of crates not being recorded.
As a solution I propose having `Cargo-minimal.lock` and `Cargo-recent.lock`. CI will just copy them over `Cargo.lock` before run.
`Cargo-minimal.lock` would contain minimal versions (See #1229 ) and would be only updated when we update `Cargo.toml` (We can probably have a check for this.) `Cargo-recent.lock` could be updated more often but always in a distinct PR to not cause merge conflicts.

### PR DESCRIPTION
**Pull request by tcharding**
This is work originally done by Kixunil in #1272, I picked it up to help out. The only changes I made were rebasingg, updating the recent lock file, adding `--locked` to hashes contrib file,  and adding a co-developed-by tag for accountability.

It could happen that we unknowingly depend on a new version of a crate without updating `Cargo.toml`. This could cause resolution issues for downstream users. It's also unclear for outsiders to know with which dependencies did we test the crate.

This change commits two lock files: `minimal` and `recent`. `minimal` contains minimal dependency versions, while `recent` contains dependency versions at the time of making the change.

Further, this adds CI jobs to test with both lock files, CI job for `internals` crate, removes old `serde` pinning and prints a warning if `recent` is no longer up to date. (We may have to override it somehow if any crate breaks MSRV.)

The documentation is also updated accordingly.

Closes #1230

### CODE REVIEW

**Review by apoelstra — DISMISSED** [review]
ACK 5a6332d5dfd4733dd0390e26de88d81c69772e99

**Comment by apoelstra** [comment]
In my local scripts, I have a comment that says "`no-std` does not work on 1.48" and I test with 1.50 instead. @tcharding that may be what you need to do here.

**Comment by apoelstra** [comment]
Neat. Apparently core2 0.3.2 works. I guess we need to set our minimal version to that.

**Comment by tcharding** [comment]
Changes in force push
- Added a patch to the front of the PR to set `core2` dependency to 0.3.2
- Updated the Cargo-minimal.lock file after building with the new `core2` dependency

**Comment by tcharding** [comment]
This is blocking `hex` crate work.

**Review by apoelstra — APPROVED** [review]
ACK c129eb6d0016ddf110c9ed72300eee53a7ca769d

**Comment by stevenroose** [comment]
I haven't followed the details of this MR, but would it be hard to have support for this in the github CI matrix thing? So specify different versions to test for certain deps and then use `update --precise` for those in the CI script?

I'm asking because in dependent crates it'd be nice to have easy access to ways to support multiple rust-bitcoin versions as well.

Ofc this would be a great tool for the rust ecosystem as a whole, but as we know I don't think many other projects care enough about msrv and dep mgmt that this kind of thing would already exist in the CI tooling.

File: contrib/test.sh [hunk 1]
```diff
@@ -2,14 +2,36 @@
...
+	done
+	if [ $dep = recent ];
+	then
+		# We always test committed dependencies but we want to warn if they could've been updated
+		cargo update
+		if diff Cargo-recent.lock Cargo.lock;
+		then
+			echo Dependencies are up to date
+		else
+			echo "::warning file=Cargo-recent.lock::Dependencies could be updated"
```
Comment (stevenroose) [inline contrib/test.sh:32]: Is this like a format that GitHub CI will pick up and somehow make visible?

**Review by stevenroose — DISMISSED** [review]
utACK e6dc6f48da7e79a3c2a188ad69ef8a9d8d799ad7

**Comment by apoelstra** [comment]
Would like an ack from @Kixunil as well since this is based on his work.

@stevenroose yeah, I think this would be generally useful. This PR does introduce CI support for this (in contrib/test.sh) though in a little bit of a hacky way. I think we should merge this and then see what kind of problems we run into before cleaning it up and maybe trying to do a more general thing. I am curious what kind of support you'd like for downstream crates....I guess, ideally your downstream projects could "merge" these lockfiles from rust-bitcoin into their own. I'm not sure if there's any tool that can do that.

File: contrib/test.sh (hunk 1 above)
Comment (tcharding) [reply contrib/test.sh:32]: This is some github syntax that gives color in their logs, @Kixunil found it but I didn't look up the docs when he told me about it.

**Review by apoelstra — CHANGES_REQUESTED** [review]
Needs rebase after #1732 since the included lockfiles now need to include honggfuzz.

BTW in #1732 I set the wrong honggfuzz version. It is at 0.5 but it needs to be 0.5.55 (or possibly lower....but certainly 0.5.0 is too low).

File: contrib/test.sh (hunk 1 above)
Comment (Kixunil) [reply contrib/test.sh:32]: IIRC it's visible when you open the lock file from the PR. It was long time since I looked into it.

**Review by Kixunil — COMMENTED** [review]
Looks reasonable, will wait until rebase with proper review. Unfortunately despite having more time I find it harder to get myself into reviewing large PRs. (Yes Tobin, it could be WOE-related, still debugging.)

**Comment by tcharding** [comment]
> (Yes Tobin, it could be WOE-related, still debugging.)

What does WOE mean, Way Of Eating?

**Comment by tcharding** [comment]
I couldn't get a minimal honggfuzz dep because where ever I put `--locked` running `./fuzz.sh` still updated the lock file. Dose the `Cargo-minial.lock` file have to commit to a minimal dependency version for honggfuzz?

**Comment by Kixunil** [comment]
> Does the `Cargo-minimal.lock` file have to commit to a minimal dependency version for honggfuzz?

Probably, but I'm not sure if we need to test with minimal dev dependencies.

**Comment by apoelstra** [comment]
@tcharding I have had no problem with a minimal version of 0.5.55.

**Comment by apoelstra** [comment]
Note that #1821 actually changes `fuzz/Cargo.toml` to have the correct minimal honggfuzz reason. So I think it's fine that this PR merely includes it in the Cargo.lock.

**Comment by tcharding** [comment]
> Note that #1821 actually changes `fuzz/Cargo.toml` to have the correct minimal honggfuzz reason. So I think it's fine that this PR merely includes it in the Cargo.lock.

Yep, sweet. I think you are right, we can merge these in any order.

File: contrib/test.sh [hunk 2]
```diff
@@ -4,12 +4,27 @@ set -ex
 
 CRATES="bitcoin hashes internals fuzz"
 
-for crate in ${CRATES}
+for dep in $deps
```
Comment (Kixunil) [inline contrib/test.sh:7]: `$deps` is not defined.

**Review by Kixunil — CHANGES_REQUESTED** [review]
Looks like there's a bug, the rest should be fine.

File: contrib/test.sh (hunk 2 above)
Comment (tcharding) [reply contrib/test.sh]: face palm, thanks man. I added `DEPS="recent minimal"`, and put quotes around `$dep` everywhere as suggested by `shellcheck`.

**Review by apoelstra — APPROVED** [review]
ACK c4c64c0dc57d31ef49d54c780b06596c711cee5a

**Review by Kixunil — APPROVED** [review]
ACK c4c64c0dc57d31ef49d54c780b06596c711cee5a

**Pull request merged**